src/driver-logitech-g300.c

```diff
@@ -0,0 +1,492 @@
+	int rc;
+
+	assert(index <= LOGITECH_G300_PROFILE_MAX);
+
+	pdata = &drv_data->profile_data[index];
+	report = &pdata->report;
+
+	switch (index) {
+	case 0: report_id = LOGITECH_G300_REPORT_ID_PROFILE_0; break;
+	case 1: report_id = LOGITECH_G300_REPORT_ID_PROFILE_1; break;
+	case 2: report_id = LOGITECH_G300_REPORT_ID_PROFILE_2; break;
+	}
+
+	rc = ratbag_hidraw_raw_request(device, report_id,
+					       (uint8_t*)report,
+					       sizeof(*report),
+					       HID_FEATURE_REPORT,
+					       HID_REQ_GET_REPORT);
+
+	if (rc < (signed)sizeof(*report)) {
+		log_error(device->ratbag,
+			  "Error while requesting profile: %d\n", rc);
+		return;
+	}
+
+	hz = logitech_g300_raw_to_frequency(report->frequency);
+
+	for (i = 0; i < profile->resolution.num_modes; i++) {
+		struct logitech_g300_resolution *res =
+			&report->dpi_levels[i];
+
+		resolution = &profile->resolution.modes[i];
+		resolution->dpi_x = res->dpi * 250;
+		resolution->dpi_y = res->dpi * 250;
+		resolution->hz = hz;
+		resolution->is_default = res->is_default;
+	}
+}
+
+static void
+logitech_g300_read_button(struct ratbag_button *button)
+{
+	const struct ratbag_button_action *action;
+	struct ratbag_profile *profile = button->profile;
+	struct ratbag_device *device = profile->device;
+	struct logitech_g300_data *drv_data = device->drv_data;
+	struct logitech_g300_profile_data *pdata;
+	struct logitech_g300_profile_report *profile_report;
+	struct logitech_g300_button *button_report;
+
+	pdata = &drv_data->profile_data[profile->index];
+	profile_report = &pdata->report;
+	button_report = &profile_report->buttons[button->index];
+
+	ratbag_button_enable_action_type(button, RATBAG_BUTTON_ACTION_TYPE_BUTTON);
+	/* Button 0 and 1 can only be mouse clicks */
+	if (button->index > 1) {
+		ratbag_button_enable_action_type(button, RATBAG_BUTTON_ACTION_TYPE_SPECIAL);
+		ratbag_button_enable_action_type(button, RATBAG_BUTTON_ACTION_TYPE_KEY);
+	}
+
+	button->type = logitech_g300_raw_to_button_type(button->index);
+
+	action = logitech_g300_raw_to_button_action(button_report->code);
+	if (action) {
+		button->action = *action;
+	}
+	else if (button_report->code == 0x00 && (button_report->modifier > 0x00 || button_report->key > 0x00))
+	{
+		struct ratbag_button_action *key_action = &button->action;
+
+		key_action->type = RATBAG_BUTTON_ACTION_TYPE_KEY;
+		key_action->action.key.key = ratbag_hidraw_get_keycode_from_keyboard_usage(
+			device, button_report->key);
+	}
+}
+
+static int
+logitech_g300_test_hidraw(struct ratbag_device *device)
+{
+	return ratbag_hidraw_has_report(device, LOGITECH_G300_REPORT_ID_GET_ACTIVE_PROFILE);
+}
+
+static int
+logitech_g300_probe(struct ratbag_device *device)
+{
+	int rc;
+	struct ratbag_profile *profile;
+	struct logitech_g300_data *drv_data = NULL;
+	int active_idx;
+
+	rc = ratbag_find_hidraw(device, logitech_g300_test_hidraw);
+	if (rc)
+		goto err;
+
+	drv_data = zalloc(sizeof(*drv_data));
+	ratbag_set_drv_data(device, drv_data);
+
```

Comment:
I meant the mouse can be either in the hardware profile mode, or in the software profile mode (where there is no profiles and the key macros are generated by LGS). So here, it would be good to force the hardware profiles.